File: src/python/zip_slip.py
"""CWE-22: Zip slip / Tar slip.

Rules: py/tarslip, py/unsafe-archive-extraction
Autofix: typically NOT provided — fix requires inserting path-canonicalization
and allow-list logic and deciding skip-vs-abort policy.
"""

import os
import tarfile
import zipfile
from flask import Flask, request

app = Flask(__name__)


@app.route("/upload-zip", methods=["POST"])
def upload_zip():
    dest = "/tmp/extract"
    os.makedirs(dest, exist_ok=True)
    with open("/tmp/uploaded.zip", "wb") as f:
        f.write(request.data)
    with zipfile.ZipFile("/tmp/uploaded.zip") as zf:
        zf.extractall(dest)
    return "ok"


@app.route("/upload-tar", methods=["POST"])
def upload_tar():
    dest = "/tmp/tar"
    os.makedirs(dest, exist_ok=True)
    with open("/tmp/uploaded.tar.gz", "wb") as f:
        f.write(request.data)
    with tarfile.open("/tmp/uploaded.tar.gz") as tf:
        tf.extractall(dest)
    return "ok"


if __name__ == "__main__":
    app.run(port=4105)

File: src/javascript/zip-slip.js
// CWE-22: Zip slip during archive extraction.
// Rule: js/zipslip
// Autofix: typically NOT provided — fix requires inserting an
// allow-list check on the resolved path against the destination root
// AND deciding what to do on violation (skip vs. abort the whole extract).

const express = require('express');
const fs = require('fs');
const path = require('path');
const yauzl = require('yauzl');

const app = express();

app.post('/extract', express.raw({ type: '*/*', limit: '20mb' }), (req, res) => {
  const dest = '/tmp/extracted';
  fs.mkdirSync(dest, { recursive: true });
  yauzl.fromBuffer(req.body, { lazyEntries: true }, (err, zip) => {
    if (err) return res.status(400).send(err.message);
    zip.readEntry();
    zip.on('entry', (entry) => {
      const target = path.join(dest, entry.fileName);
      if (/\/$/.test(entry.fileName)) {
        fs.mkdirSync(target, { recursive: true });
        zip.readEntry();
      } else {
        zip.openReadStream(entry, (e, rs) => {
          if (e) return res.status(500).send(e.message);
          fs.mkdirSync(path.dirname(target), { recursive: true });
          rs.pipe(fs.createWriteStream(target)).on('close', () => zip.readEntry());
        });
      }
    });
    zip.on('end', () => res.send('ok'));
  });
});

app.listen(3105);
